TOOLS.rst

```diff
@@ -0,0 +1,201 @@
+This file explains the development tool chain around ixdat
+
+Tools used
+==========
+
+The following is a list of specific tools used during ixdat
+development:
+
+* **pytest** is used for running tests
+* **flake8** is used for linting
+* **sphinx** is used to build documentation
+
+The following is a list of "tool and commands runners":
+   
+* **invoke** is used during development, to run tools and other
+  pre-configured maintenance tasks inside the existing development
+  environment
+* **tox** is used as a tool to run all **QA** tools and commands
+  across all supported enviroments, typically before push or by
+  continuous integration (CI) tools
+
+Tool and command runners in more detail
+=======================================
+
+The reason there are two different tool runners is that they have very
+different purposes:
+
+**invoke** is used as general purpose, platform independent way of
+defining and running "tasks relevant for development" (reminiscent of
+the *makefiles* of olden times, commonly used for software development
+on Unix systems). These tasks can be something like running a linter,
+type checker, test runner or code formatter, building documentation,
+cleaning the repository of temporary files, building a package or
+updating the list of contributors from the git log.
+
+**tox** is tuned towards a specific purpose, which is to run
+specifically QA tools (something like linter, type checker, test
+runner or code formatter) in a clean environment, with a fresh install
+of code, across as many supported python versions as possible. ``tox``
+can also be used in CI tools, such a Github Actions, to specify what
+to run as part of the CI, which eliminates the need to an extra
+configuration.
+
+Both tools share the purpose of hiding the details of the way the
+tools should be run (arguments, settings etc.) and defining a single
+place to define those.
+
+From the definitions above, however, it is obvious that the QA tools
+may be run by both. This is ok, because the purposes are different.
+But it raises the important concern of *where* the one true place to
+define the arguments and settings are for those. See definitions of
+that in the following section.
+
+Where are settings and arguments kept
+-------------------------------------
+
+For settings and arguments for tools the following rules apply:
+
+* **setup.cfg** If the tool support defining them in the shared
+  configuration file ``setup.cfg``, then they belong there no-matter
+  how they are run. If not (settings may only be available at the
+  command line), see the next points.
+* **tox.ini** If the tool is run by both ``tox`` and ``invoke``, then
```

Comment:
Maybe I didn't formulate it clearly. If the tool support being configured in setup.cfg, then it is configured there, always. setup.cfg is an attempt to clean up the mess of ``.something_rc`` files that used to live in the root of projects, because they all had their own configuration files. Unfortunately, not all tools support it, but quite many does, so we gather those.

If, and only if, the tool doesn't support configuration via setup.cfg (or its own configuration file), but only allows configuration on the command line, then we move the actual settings into tox.ini or tasks.py. And of course the actual name of the command itself will always have to be listed in one of the these two files.

As for the **tox precedence**, that is unfortunately the only way that it can be, since tox' configuration is just an ini-file and it can't read command lines from tasks.py, but seeing as tasks.py is a real python-file, we can do it the other way. But keep in mind that the precedence that you mentioned is only ever relevant/present for **QA tools**, of which I have already configured 2 (test-runner, linter) out of the 4 (test runner, linter, code format checker and possibly static type checker) that one might usually have and I plan to add the 3 also (code format checker), so most of this shared configuration is already done.

Also keep in the mind that this is relevant **only for the developer who wants to add or tweak a tool**. It is totally invisible for developers who wants to just use it. An ordinary developer who wants to take the dive and start using the tools and tasks need know only one command **invoke** (just like make in the old days) and she will ask invoke for everything she want's to know about what she can do (or consult the last sections of "Getting started developing", which will also only mention invoke commands). That is in fact the whole point about invoke, to make it easier for new developers by making it so that they do not need to know the names and common arguments of all the tools.

**tox** becomes relevant only for "top" developers or maintainers, who might want to run the tests in all the pythons they have available, to check cross python version compatibility locally maybe before a release. And it is used by the CI platforms such as Github Actions (which was one of the large motivators for setting tox up in the first place). And it is used by the commit hooks purely because it makes it possible to run all the QA in parallel which makes it run quite a bit faster, which you really want if you are waiting to see your push finish.

Does this alleviate your concern?

Reply:
Thanks a lot for this. I am happy about the different roles of invoke and tox, and the relationship between the two.

I however, after thinkng more and understanding it better, prefer that they don't share settings, due to the complexity of the implementation (and its explanation in this .rst), and my argument here: https://github.com/ixdat/ixdat/pull/3#discussion_r556428028

Reply:
Done, as per a comment above. Of course I forgot to update the TOOLS.rst to reflect the change. I will force push a amended commit in a second.

Reply:
With that change the section in TOOLS.rst becomes: